pyk/src/pyk/cterm/symbolic.py

```diff
@@ -125,12 +150,18 @@ def execute(
 
         state = CTerm.from_kast(self.kore_to_kast(response.state.kore))
         resp_next_states = response.next_states or ()
-        next_states = tuple(
-            NextState(
-                CTerm.from_kast(self.kore_to_kast(ns.kore)),
-                self.kore_to_kast(ns.rule_predicate) if ns.rule_predicate is not None else None,
+        branching_constraints = tuple(
+            self.kore_to_kast(not_none(s.rule_predicate)) if s.rule_predicate is not None else None
+            for s in resp_next_states
+        )
+        # Branch constraint minimization makes sense only if there is a proper branching
+        if len(branching_constraints) >= 2 and all(bc is not None for bc in branching_constraints):
+            branching_constraints = self.minimize_constraints(
```

Comment:
A few questions:

1. Is removing the extra constraints necessary for proving, or just an aesthetic thing?
2. Do we understand how these spurious constraints are introduced?
3. How general is this solution? I'd like to avoid the situation that we keep extending this class with new herusitics to handle spurious constraints that we might run into.
4. Can this be done on the backend more efficiently?

Reply:
1) Removing the extra constraints is not necessary per se for proving, but it is important, for two reasons:
    a) "objectively", these constraints are not real branching constraints, and the intended meaning of the `rule_predicate` field, coming from the backend, is to only contain real branching constraints - I believe that the backend team is in agreement on this (@jberthold?);
    b) subjectively, as part of the user experience of the person doing the proofs, I have found it important to know precisely what we are branching on, and I found these additional constraints both confusing and non-informative.
2) Yes, we understand this. They are introduced because the backend, when branching, only checks satisfiability of obtained constraints given the path condition, and does not check if there are some constraints coming from the `requires` clauses of the rules on which the branching happens that are entailed by the path condition using Z3, but only using the simplifications at hand.
3) I believe that this solution will cover the great majority of the cases we will encounter, and can't think right now of reasonable scenarios in which it wouldn't. For one, I don't see how there would be a constraint that is in the intersection that is not entailed by the path condition. @jberthold, perhaps you have some more thoughts on this?
4) @jberthold, what do you think? I think that, since this is a heuristic, it could be the case that it fits `pyk` better. The complexity of the procedure would be the same - find intersection, call SMT solver to check entailment. However, given 1a) above, perhaps this reasoning should indeed go to the backend.

Reply:
> 1. ...the intended meaning of the `rule_predicate` field, coming from the backend, is to only contain real branching constraints - I believe that the backend team is in agreement on this (@jberthold?);
>
> 4. @jberthold, what do you think? I think that, since this is a heuristic, it could be the case that it fits `pyk` better. The complexity of the procedure would be the same - find intersection, call SMT solver to check entailment. However, given 1a) above, perhaps this reasoning should indeed go to the backend.

We could probably implement this filtering in the backend, but when we discussed the open github issue, we concluded that the frontend would be a better place to implement it.
Adding such a filter changes the understanding of what the branching condition is. At the moment, the backend returns, for each branch of a branching result, _all_ constraints that were _added_ on that branch (from a rule's side conditions. We are looking specifically at branches that arise from conditional rules where the condition is not determined)). As we found in the case at hand, some of the constraints are redundant but they are the basis of the state split that  happens in the branch.

> 3. ... I don't see how there would be a constraint that is in the intersection that is not entailed by the path condition. @jberthold, perhaps you have some more thoughts on this?

I don't have a ready-made example, and I imagine this to not be a useful scenario in practice, but I could imagine that one can construct a case where an execution branches into, say 3 possible next states, but then one of those 3 states is filtered out because of other, unrelated, path conditions. The two remaining next states of the branch may include a branching condition that is not entailed yet (i.e., constitutes new knowledge about the state).

Reply:
I've talked with @ehildenb about this as well and he also thinks `pyk` is the correct place, for now, for this filtering.